### PR TITLE
New timetrack

### PR DESCRIPTION
**Pull request by rodolforg**
I don't know all features of current TimeTrack, please let me know if something is missing:

- [x] Sync with Parameters dock (it has issues with two timetracks enabled. It works fine with only the new timetrack.

- [x] context/popup menu via right-click in waypoint (for single waypoint)

- [x] move selected waypoint (mouse drag and arrow keys)

- [x] shift + drag : copy selected waypoint
- [x] waypoint deletion via Delete key (**novelty**)

- [x] double-click opens Waypoint editor dialog

- [x] multiple selection - including across parameters (**novelty**) (#761 #762)

- [x] display static interval (**novelty**) (#790)

- [x] allow to scale waypoints (**novelty**) (#266 https://github.com/synfig/synfig/issues/266#issuecomment-480610390 )

- [x] parameters and Timeline panels show the same row selected. (**novelty**) (#246)

- [x] waypoint navigation  (**novelty**)
  key `n` and `b` to go to next or previous waypoint. By holding `shift`, user skips 5 waypoints.
(by design, it only works when only one waypoint is selected.)

- [ ] #529  fixed?

### CODE REVIEW

**Comment by ice0** [comment]
Wow!

**Comment by morevnaproject** [comment]
!!! O___O

**Comment by morevnaproject** [comment]
At the moment I am finishing release of Synfig 1.3.12, and then I need to finalize our GSoC submission. Then I will review this. Thank you so much!!! ^__^

**Comment by morevnaproject** [comment]
/remind me about this in 5 days

**Comment by rodolforg** [comment]
> In the screenshot below (built from synfig master, not rodolforg version) the keyframes (start and end) don't appear fully making it hard for me to duplicate the keyframes if I wanted to. Hope I read this [commit message](https://github.com/synfig/synfig/pull/1135/commits/56078c3903148e5879bf02fe27cb951d582747f9 .

Hm… Sadly I just 'replicate' the current behavior of 2px displacement the current Timetrack has.

**Comment by ice0** [comment]
As far as I remember, @morevnaproject also has some problems in this part.

**Comment by morevnaproject** [comment]
Yes adding padding at the start and end will make life easier for many users ^__^ - #1145

**Comment by morevnaproject** [comment]
I am building this PR now... ^__^

**Comment by morevnaproject** [comment]
Tested and it is so awesome!

So far i found only one issue:
If i select any waypoints and hit Delete, then Timetrack is not updated. The animation in workarea updated - so I can see that waypoints are really deleted. To make changes reflected on Timetrack I need to click with on any place of Timetrack.

**Comment by rodolforg** [comment]
> If i select any waypoints and hit Delete, then Timetrack is not updated. 

:o Are you sure? I know this phrase sucks but: "Here it works as expected"

**Comment by morevnaproject** [comment]
Yes, here is a video ^__^ - https://www.youtube.com/watch?v=b6-2QW6PWds

Built with commit 188ac4a87edc40 merged into master at 00e3c96c42a7566.

**Comment by morevnaproject** [comment]
I have built commit 188ac4 (without merging into master) and still can reproduce the same issue. ^__^"

**Comment by rodolforg** [comment]
OMG, I'll have to check it out because here is my video https://youtu.be/6E1JXP8UuK4

**Comment by morevnaproject** [comment]
Could it be because of difference in GTK versions? I am using Gtk 3.22.30.

BTW, I love semi-transparent waypoints - didn't discovered them at first! ^__^

**Comment by rodolforg** [comment]
> Could it be because of difference in GTK versions? I am using Gtk 3.22.30.

Here I am with 3.24.11.

> BTW, I love semi-transparent waypoints - didn't discovered them at first! ^__^
They're like that for valuenodes, just like I did for Curves Dock.

File: synfig-studio/src/gui/widgets/widget_timetrack.cpp [hunk 1]
```diff
@@ -0,0 +1,825 @@
...
+			param_list.add("addcanvas", value_desc.get_value().get(synfig::Canvas::Handle()));
+		} else {
+			param_list.add("addvaluedesc", value_desc);
+		}
+
+		param_list.add("addtime", wi->time_point.get_time());
+	}
+	action->set_param_list(param_list);
+	canvas_interface->get_instance()->perform_action(action);
+}
```
Comment (rodolforg) [inline synfig-studio/src/gui/widgets/widget_timetrack.cpp]: @morevnaproject , could you try if a `queue_draw();` here solves your issue?

Comment (morevnaproject) [reply synfig-studio/src/gui/widgets/widget_timetrack.cpp]: This solves issue, but please see my next comment...

**Comment by morevnaproject** [comment]
Hi! Apologies for dealy.
I have made more testing and discovered some interesting things:
1. Timetrack is also not updates when new waypoints are created. I do not know why I didn't discovered that initially - maybe I had other (old) Timetrack Panel active while animating. Or maybe because of next thing (2).
2. The issue with timetrack not updating is happens ONLY for new documents. If I open some existing document, then timetrack updates correctly (both for creating and deleting waypoints).

Please see this video for demonstration ^__^ - https://www.youtube.com/watch?v=_wwFc3nYwcc

**Comment by morevnaproject** [comment]
P.S. All tests are done for 188ac4a

**Comment by rodolforg** [comment]
> Hi! Apologies for dealy.
Me too! :)

>     2. The issue with timetrack not updating is happens ONLY for new documents. If I open some existing document, then timetrack updates correctly (both for creating and deleting waypoints).

Ok, I investigate it more and I found out when exactly it isn't updated: when the parameter is not a valuenode before you select layer.
* If you select a circle layer whose Origin is animated, and edit its animation it refreshs the timetrack.
* If you then starts to animate its radius, it doesn't visually refresh its waypoints in timetrack.
* Then you select none, select circle layer back and continue to animate its radius: everything works as expected.

The reason for it is that I set to only redraw the timetrack when a valuenode changes. I choose to do like that due to performance reasons (if it has a large parameter or waypoint list, maybe timetrack would be slowly re-rendered when user simply changes a not-animated parameter).

The problem is I only search for value nodes in the parameter list when it is set. I do not track if a regular value base changes to value node - because I don't know how to do it yet. If I can't find out when editor changes a regular parameter to an animated value node, ... well, I'll have to refresh everytime anything changes.

**Comment by rodolforg** [comment]
@morevnaproject I think that issue you pointed is now fixed. I made a workaround and it seems to work fine.
The only thing left is… sync the Param panel height with TimeTrack Panel (last row issue).
I hope I solve it today.

**Comment by rodolforg** [comment]
Not that "today", but it's done :)

I believe this PR is ready now and there's nothing left to do here. :)
Please tell if there is any bug @morevnaproject @ice0 @Jose-Moreno

File: synfig-studio/src/gui/widgets/widget_timetrack.cpp (hunk 1 above)
Comment (rodolforg) [reply synfig-studio/src/gui/widgets/widget_timetrack.cpp]: I believe it's solved now.

**Comment by rodolforg** [comment]
> I believe this PR is ready now and there's nothing left to do here. :)

Hm... Actually I forgot about this: https://github.com/synfig/synfig/issues/266#issuecomment-573763300 Should I do it?

And I found out issue #529 , and I found out my TimeTrack widget make Synfig crash. I'll fix it now

**Comment by morevnaproject** [comment]
Hi! Thank you for the fixes! Previous issue is resolved.

I have tested more and found another issue:

1. Create new document.
2. Create circle, animate any of its parameters.
3. Hit Ctrl+S -> "Save as" dialog appears.
4. Close that dialog.
5. After that you will be unable to move waypoints.

Demonstration video -  https://youtu.be/NUJ2Hqpm0SM?t=95 (starting from 1:35)

> Hm... Actually I forgot about this: #266 (comment) Should I do it?

I think it would be nice to have panel with buttons, so user do not need to guess how to scale waypoints. Of course, it not needs "bounding box selection" button, because you implemented it in a very intuitive way. ^__^

**Comment by rodolforg** [comment]
> And I found out issue #529 , and I found out my TimeTrack widget make Synfig crash. I'll fix it now
fixed

**Comment by rodolforg** [comment]
>     5. After that you will be unable to move waypoints.
> Demonstration video - https://youtu.be/NUJ2Hqpm0SM?t=95 (starting from 1:35)

Fixed :)

> > Hm... Actually I forgot about this: #266 (comment) Should I do it?
> 
> I think it would be nice to have panel with buttons, so user do not need to guess how to scale waypoints.

My only concern is about cluttering the dock, reducing its size.

@morevnaproject  and could please check if #529 is addressed?

**Comment by morevnaproject** [comment]
>>    Demonstration video - https://youtu.be/NUJ2Hqpm0SM?t=95 (starting from 1:35)
>
> Fixed :)

Confirmed! Thank you! ^__^

> My only concern is about cluttering the dock, reducing its size.

I think that's okay as long as toolbar is on the left (we have a plenty of horizontal space, while vertical space is usually limited). ^__^

**Comment by morevnaproject** [comment]
Regarding #529 - it is not fixed. Well, actually... maybe it  is partially fixed.
I wrote detailed explanation in comment here - https://github.com/synfig/synfig/issues/529#issuecomment-594645256

It is possible this problem with updating timetrack might have deep roots and might be hard to fix, so, maybe it can be solved later. This is why I do not put explanations in this PR and instead placed them as comment into #529. ^__^

File: synfig-studio/src/gui/docks/dock_timetrack2.cpp [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+#include <gui/localization.h>
+
+#include <synfig/general.h>
+
+#endif
+
+using namespace studio;
+
+Dock_Timetrack2::Dock_Timetrack2()
+	: Dock_CanvasSpecific("timetrack2", _("Timetrack"), Gtk::StockID("synfig-timetrack")),
```
Comment (morevnaproject) [inline synfig-studio/src/gui/docks/dock_timetrack2.cpp]: I suggest to change into this:
```
Dock_CanvasSpecific("timetrack", _("Timetrack"), Gtk::StockID("synfig-timetrack")),
```

...and change initialization of old timetrack (in `dock_timetrack.cpp`) to this:
```
Dock_CanvasSpecific("timetrack-old", _("Timetrack (old)"), Gtk::StockID("synfig-timetrack")),
```

File: synfig-studio/src/gui/app.cpp [hunk 3]
```diff
@@ -1070,6 +1072,7 @@ DEFINE_ACTION("panel-curves",          _("Graphs"));
 DEFINE_ACTION("panel-groups",          _("Sets"));
 DEFINE_ACTION("panel-pal_edit",        _("Palette Editor"));
 DEFINE_ACTION("panel-soundwave",       _("Sound"));
+DEFINE_ACTION("panel-timetrack2",      _("Timetrack 2"));
```
Comment (morevnaproject) [inline synfig-studio/src/gui/app.cpp]: change to:
```
DEFINE_ACTION("panel-timetrack-old", _("Timetrack (old)"));
DEFINE_ACTION("panel-curves",          _("Graphs"));
DEFINE_ACTION("panel-groups",          _("Sets"));
DEFINE_ACTION("panel-pal_edit",        _("Palette Editor"));
DEFINE_ACTION("panel-soundwave",       _("Sound"));
DEFINE_ACTION("panel-timetrack",      _("Timetrack"));
```

File: synfig-studio/src/gui/app.cpp [hunk 4]
```diff
@@ -1240,6 +1243,7 @@ DEFINE_ACTION("keyframe-properties", _("Properties"));
 "		<menuitem action='panel-info' />"
 "		<menuitem action='panel-navigator' />"
 "		<menuitem action='panel-timetrack' />"
+"		<menuitem action='panel-timetrack2' />"
```
Comment (morevnaproject) [inline synfig-studio/src/gui/app.cpp]: change to:
```
"		<menuitem action='panel-timetrack-old' />"
"		<menuitem action='panel-timetrack' />"
```

**Comment by morevnaproject** [comment]
In general I feel this PR is ready for merge. 

I just suggest to change 
"timetrack" -> "timetrack-old"
and 
"timetrack2" -> "timetrack".
(see comments in my review).

After doing those replaces the users will get new timetrack instead of old, without changing anything in their configuration.

I think it is not needed to rename .cpp/.h files, as this will help to keep track of version history.

**Comment by morevnaproject** [comment]
...and one more thing: I found that semi-transparent representation of waypoints makes users think that they are "inactive" (cannot be moved).
screenshot_005

Can we think about some different representation? Maybe remove transparency and have a double outline for them?

**Comment by rodolforg** [comment]
> I think it is not needed to rename .cpp/.h files, as this will help to keep track of version history.

We can use "git mv" to rename files and its history log is preserved.

**Comment by rodolforg** [comment]
Should I rename the classes too? eg. `Dock_Timetrack` to `Dock_TimetrackOld`

File: synfig-studio/src/gui/app.cpp (hunk 4 above)
Comment (rodolforg) [reply synfig-studio/src/gui/app.cpp]: Done

File: synfig-studio/src/gui/app.cpp (hunk 3 above)
Comment (rodolforg) [reply synfig-studio/src/gui/app.cpp]: Done

File: synfig-studio/src/gui/docks/dock_timetrack2.cpp (hunk 2 above)
Comment (rodolforg) [reply synfig-studio/src/gui/docks/dock_timetrack2.cpp]: Done

**Comment by rodolforg** [comment]
> Can we think about some different representation? Maybe remove transparency and have a double outline for them?

Or maybe a smaller waypoint?

**Comment by rodolforg** [comment]
> I think that's okay as long as toolbar is on the left (we have a plenty of horizontal space, while vertical space is usually limited). ^__^

The button panel should be left or right timetrack widget? :)
Scale and Move buttons only?

**Comment by morevnaproject** [comment]
> We can use "git mv" to rename files and its history log is preserved.

Well, sometimes I am filtering git history "by filename". I am not sure how this will be handled.
In any case, I am not against of renaming files, and I agree that it will be a cleaner solution. ^__^

> Should I rename the classes too? eg. Dock_Timetrack to Dock_TimetrackOld

That's good idea too! ^__^

>> Can we think about some different representation? Maybe remove transparency and have a double outline for them?
> Or maybe a smaller waypoint?

I still think about double outline, because this is more similar to representing "multiple waypoints" (for my taste ^__^). 

> The button panel should be left or right timetrack widget? :)

On the right, sorry for the mess. ^__^

> Scale and Move buttons only?

I can't think about any other buttons right now...

Well actually, we can have a buttons with each interpolation type. Clicking such button will apply this interpolation to selected waypoints. That would be cool. But I think it is better to leave for separate PR - I really would like to have this PR merged as soon as possible. ^__^

**Comment by rodolforg** [comment]
> In any case, I am not against of renaming files, and I agree that it will be a cleaner solution. ^__^

Ok. I tested `git log`, and `--follow` is needed to follow file renaming. However, it doesn't work for `git blame` unless you specify previous commit :P So, I'm letting this as it is now.

> > Should I rename the classes too? eg. Dock_Timetrack to Dock_TimetrackOld
> 
> That's good idea too! ^__^

Done.

> > > Can we think about some different representation? Maybe remove transparency and have a double outline for them?
> > > Or maybe a smaller waypoint?
> 
> I still think about double outline, because this is more similar to representing "multiple waypoints" (for my taste ^__^).

Ok. I'll try it.

Hm... Actually, regular line width is 1.0; selected waypoints have width of 2.0.
So, linked-waypoints (not necessarily it means multiple waypoints) should use 2.0 and 4.0?

> > The button panel should be left or right timetrack widget? :)
> 
> On the right, sorry for the mess. ^__^
> 
> > Scale and Move buttons only?
> 
> I can't think about any other buttons right now...
> 
> Well actually, we can have a buttons with each interpolation type. Clicking such button will apply this interpolation to selected waypoints. That would be cool. But I think it is better to leave for separate PR - I really would like to have this PR merged as soon as possible. ^__^

Ok. Move, Duplicate and Scale.

**Comment by ice0** [comment]
> Hm... Actually, regular line width is 1.0; selected waypoints have width of 2.0.
> So, linked-waypoints (not necessarily it means multiple waypoints) should use 2.0 and 4.0?

If linked waypoints can also be selected or not, then using width, in my opinion, is not a good idea. Better to change the color.

**Comment by rodolforg** [comment]
> If linked waypoints can also be selected or not, then using width, in my opinion, is not a good idea. Better to change the color.

Sorry, my mistake: I mean converted waypoints. (anyway, it doesn't necessarily mean "multiple waypoints").

**Comment by rodolforg** [comment]
@morevnaproject  sorry for the long time I spent on this.
Could you please test and evaluate it?
And I still don't know what to do with converted waypoint design…

**Comment by morevnaproject** [comment]
Hi guys! Apologies for delay. Let me clarify what I was meaning with "double outline" ^__^ -
screenshot_001

**Comment by morevnaproject** [comment]
Tested the PR. The toolbar on the right looks so cool! ^__^
screenshot_001

I suggest to merge this PR already and do all further improvements as separate PRs.

@rodolforg can you please rebase this branch on top of the latest master? ^__^

**Comment by morevnaproject** [comment]
Merged now. Thank you so much!!! :tada:

**Comment by morevnaproject** [comment]
I have added two issues for future improving of TimeTrack:
- https://github.com/synfig/synfig/issues/1266
- https://github.com/synfig/synfig/issues/1267

**Comment by rodolforg** [comment]
@morevnaproject Could you please close the issues mentioned on this PR description?
If they're solved, of course.

**Pull request merged**